pep-9999.rst

```diff
@@ -0,0 +1,252 @@
+that are constructed with ``covariant=True`` or ``contravariant=True``, 
+e.g.:
+::
+
+    from typing import TypeVar, Generic, Protocol
+    
+    T_co = TypeVar("T_co", covariant=True)
+    T_contra = TypeVar("T_contra", contravariant=True)
+    
+    class SimpleFunction(Protocol[T_contra, T_co]):
+        def __call__(self, x: T_contra) -> T_co: ...
+
+But type variables can also be used outside of the scope of generic 
+types, where the concept of variance does not apply, e.g.:
+::
+
+    def first(things: Iterable[T]) -> T:
+        return next(iter(things))
+
+
+This example is valid, but using ``T_co`` or ``T_contra`` instead of 
+``T`` here does not make sense, and type checkers do not allow is.
+
+In ``SimpleFunction``, ``T_co`` or ``T_contra`` are used to annotate
+its method as well. This is required because the variance is associated 
+with the individual type variables. But the method signature is unrelated 
+to the variance of the type parameters of its generic type, 
+``SimpleFunction`` in this example. 
+
+Because variance is restricted to the type parameters of generic types, 
+it makes more sense to specify the variance of type parameters directly
+where the generic type is defined, removing the need to specify it on 
+the type variables themselves. 
+
+
+Proposal
+========
+
+The variance of the type parameters of a generic class can be specified
+with the ``+`` and ``-`` prefix operators on their respective type
+variables within the type parameter list of ``typing.Generic``, 
+``typing.Protocol``, or the generic base class. Instead of
+::
+
+    T_co = typing.TypeVar('T_co', covariant=True)
+    T_contra = typing.TypeVar('T_contra', contravariant=True)
+
+    class SupportsInvertOld(typing.Protocol[T_co]):
+        def __invert__(self) -> T_co: ...
+
+    class SupportsPartialOrderOld(typing.Protocol[T_contra]):
+        def __le__(self, other: T_contra): ...
+
+
+The new syntax uses the ``+`` and ``-`` prefix operators to specify
+variadic type parameters in the same place where the generic type is
+declared:
+::
+
+    T = TypeVar("T")
+    
+    # covariant
+    class SupportsInvert(typing.Protocol[+T]):
+        def __invert__(self) -> T: ...
+
+    # contravariant
+    class SupportsPartialOrder(typing.Protocol[-T]):
+        def __le__(self, other: T): ...
+
+This syntax is inspired by Scala programming language [1]_.
+
+
+
+Specification
+=============
+
+
+Valid use locations
+-------------------
+
+
+The new type variable variance syntax can only be used within the type
+parameter list of ``typing.Generic``, ``typing.Protocol``, or a generic 
+base class. 
+
+The ``+`` and ``-`` prefix operators can be used on ``typing.TypeVar``
+and ``typing.ParamSpec`` instances, that are now already variadic, i.e.
+its ``__covariant__`` and ``__contravariant__`` attributes must be 
+``False``.
+
+When the same type variable is used on multiple generic base classes,
```

Comment:
I guess the same constraint applies for the case of a single generic base class. E.g. the following should be rejected by the type checker

```
class A(Generic[T]): ...  # Invariant in T.
class B(A[+T]): ... 
```

because a covariant type parameter `T` is used in an invariant position in `A[...]`.

Reply:
I'm not sure that this is incorrect. At least, mypy allows it, and behaves as expected here:

```python
from typing import TypeVar, Generic

T = TypeVar('T')
T_co = TypeVar('T_co', covariant=True)
T_contra = TypeVar('T_contra', contravariant=True)


class Box(Generic[T]):
    value: T

    def __init__(self, value: T):
        self.value = value


class BoxCo(Box[T_co]): ...
class BoxContra(BoxCo[T_contra]): ...


class A: ...
class B(A): ...


def accepts_a(x: Box[A]) -> None: ...
def accepts_a_co(x: BoxCo[A]) -> None: ...
def accepts_a_contra(x: BoxContra[A]) -> None: ...
def accepts_b(x: Box[B]) -> None: ...
def accepts_b_co(x: BoxCo[B]) -> None: ...
def accepts_b_contra(x: BoxContra[B]) -> None: ...


a_box: Box[A] = Box(A())
a_box_co: BoxCo[A] = BoxCo(A())
a_box_contra: BoxContra[A] = BoxContra(A())

b_box: Box[B] = Box(B())
b_box_co: BoxCo[B] = BoxCo(B())
b_box_contra: BoxContra[B] = BoxContra(B())

accepts_a(a_box)
accepts_a_co(a_box_co)
accepts_a_contra(a_box_contra)

accepts_a(b_box)  # incompatible type "Box[B]"; expected "Box[A]"
accepts_a_co(b_box_co)
accepts_a_contra(b_box_contra)  # incompatible type "BoxContra[B]"; expected "BoxContra[A]"

accepts_b(a_box)  # incompatible type "Box[A]"; expected "Box[B]"
accepts_b_co(a_box_co)  # incompatible type "BoxCo[A]"; expected "BoxCo[B]"
accepts_b_contra(a_box_contra) 

accepts_b(b_box)
accepts_b_co(b_box_co)
accepts_b_contra(b_box_contra)
```

My guess is that defining `T` as covariant does not change the variance of `T` in A, i.e. even though the same type variable `T` is used to define the type parameters to `A` and `B`, the type parameters are independent from one another (not considering type constraints and bounds). 

But I can see that this could lead to issues, and that this should be discussed. Although I don't think that this PEP is the best place to impose additional restrictions on generic types, but should be left to type checkers, since type variance is not a concept that is specific to Python.

Reply:
You're right, it is not as straightforward as I thought initially. 

I put together an example [demonstrating](https://mypy-play.net/?mypy=latest&python=3.10&gist=80be2a9951f8eff41f2a9ea3c2e1beee) unsoundness in my earlier comment -- the value of type `A[int]` is also typed as `A[bool]` in the same program, even though `A` is invariant in `T`. 

I would much prefer to settle this in a PEP instead of leaving it open to interpretation by tooling authors. Happy to start a thread on typing-sig@ if you feel this is best addressed in a separate PEP.